### ISSUE TITLE
[feature request] nixos/initrd-ssh: run a Tor hidden service

### ISSUE THREAD

**Issue by volth** [NixOS/nixpkgs#35900]
## Issue description

There is a ```initrd-ssh``` in ```NixOS``` which runs ```dropbear``` in ```initrd``` allowing to log in to the machine on the  earliest stages of the boot process (to, for example, enter ```cryptsetup``` password and mount encrypted root partition).

There are two problems though - the machines with dynamic IP and the machines behind NAT.

It would be nice to run a Tor hidden service on ```initrd``` to get access to those machines.

**Comment by oxij** [comment]
I'm sure it can be done today with some ugly hackery (you need to pack service private keys and tor db into initrd, ugh).

A proper implementation requires initrd to become "a very small nixos" and proper bootloader handling (see #33940 for discussion).

**Comment by volth** [comment]
Would that "very small nixos" include systemd? (not for trolling, just because the topic was ignored in #33940).

**Comment by oxij** [comment]
You can do that with systemd (or with another, superior, init system like s6), but you can also just make services less systemd-specific and just `toposort` their exec scripts. But some very core contributor were against this the last time that was discussed (see #11484 and follow ups). I implemented most of that around #2920, but when the best parts of #2920 got rejected I stopped bothering to PR the rest (until #11484, that is, because it touched fileSystems, not just initrd).

See also the "The Church of Suckless NixOS is looking for followers" and "Nix init system" ML threads for more discussion on this.

I think if you wait long enough you'll get this feature in NixOS because _everything_ I ever did for NixOS initrd that was rejected then got reimplemented (usually, in an inferior way) after ~3-5 years. SLNOS got some awesome generic service infra recently, so see you in 3-5 years, I guess :)

**Comment by volth** [comment]
Even without Tor, there are already two "nixos services" in initrd (networking and dropbear).
They are not exactly correspond to nixos (networking is much basic), but yes, as the number of initrd services grows (Tor, Nitrokey, what else?) they could form a micro-nixos

**Comment by oxij** [comment]
Right. Like I said, unless some core maintainer gets the message and starts accepting patches that move nixos towards a proper solution people will just keep adding ad-hoc services to initrd and nixos will end up with two sets of expressions for all the core services.

I bet NTP would be the next big thing because a lot of services require precise time. Neither of tor, i2p, cjdns, or gnunet will work without precise time. I also expect more of `networking` to get ported to `initrd.network`.

Then, let me play a prophet here, the duplication of work situation will get so bad that nixpkgs community will organize to resolve the problem by inventing another crazy mechanism (like another flag to each core service that requests its presence in initrd, or something), see `isContainer` flag and #33922 (which, IMHO, didn't get merged only because it makes the craziness of `isContainer` apparent). Note that I'm not criticizing any of `isContainer` work or any other real code/PRs I mention, all of them are the best technical solutions for constraints nixos core maintainers place on accepted PRs. I'm criticizing the constraints. nixpkgs' policies suck. See around https://github.com/NixOS/nixpkgs/issues/36229#issuecomment-370218978 for more discussion.

Again. A proper solution?

1) Get rid of hard dependency on systemd (see "Modifying the init system (introducing S6 supervision suite)" thread on ML for more details).
   a) Make services generate generic "jobs" (btw, nixos had that before systemd coup) instead of systemd-specific things.
   b) You can then replace the underlying init system with something more sane (like s6), if you wish.
   c) Make services monoidal.
   d) Use topological sorting everywhere, but offload it to init system when it supports it. (Sounds more complicated that it actually is.)
2) Make `initrd` into an instance of `nixos`.

As a free bonus:

- Make containers into an instances of `nixos`, get rid of `isContainer`.
- Make ISOs a simple wrapper that carries several instances of `nixos`, get rid of the abomination you have for ISO generation now.
- Make a proper netboot module that can carry several instances of `nixos` (actually, you can do that today in a patchset that applies to mainline nixos, because nixos doesn't yet implement most of #2920 in any crazy-ass way, but I'm unwilling to PR it for aforementioned policy reasons).

Is the proper solution likely to get merged? No. Hence, see you in 3-5 years, when initrd gets enough networking and NTP to run Tor.

I'm reasonably sure SLNOS will go public before that. Just sayin'

**Comment by oxij** [comment]
Also /cc @esoeylemez and @7c6f434c since esoeylemez is the original author of the monoidal services proposal and 7c6f434c's https://github.com/7c6f434c/7c6f434c-configurations were another big part of the inspiration.

The comment above and https://github.com/NixOS/nixpkgs/pull/34645#issuecomment-367058041 describes the current state of things in SLNOS.

Feel free to ask any questions or provide any suggestions. (I don't know when it'll go public yet, I'm sorry.)

**Comment by edolstra** [comment]
@oxij This issue tracker is intended for improving NixOS, not advertising other projects.

IMHO the best solution would be to run systemd in the initrd. @lethalman did some work on that a couple of years ago but got stuck at some point (https://nixos.org/nix-dev/2014-October/014802.html).

**Comment by oxij** [comment]
> @oxij This issue tracker is intended for improving NixOS, not advertising other projects.

First, it's not advertising, since there's no product to advertise (yet), and no commercial interest involved (ever).

Secondly, I'm sorry, I would love to improve NixOS, but when NixOS doesn't want to merge said improvements and then there's a discussion that pretty much specifically asks for those improvements I think I'm not breaching any ethical barriers here by trying to keep the discussion of the proper solution going and nudging the community to adopt it.

Third, I hope one day sanity prevails and NixOS will either merge monoidal services from SLNOS or just reimplements them by itself, and then SLNOS will become a marginally oriented set of security patches.

Which part of this is not about "improving NixOS"?

**Comment by 7c6f434c** [comment]
@edolstra weren't you in favour that some breaking experiments are top be run in external branches anyway?
